collections/src/main/scala/strawman/collection/mutable/Queue.scala

```diff
@@ -29,46 +29,49 @@ import scala.{Int, Boolean, Unit, Option, Some, None, NoSuchElementException, Se
   *  @define willNotTerminateInf
   */
 @SerialVersionUID(3L)
-class Queue[A]
-  extends MutableList[A]
-    with LinearSeqOps[A, Queue, Queue[A]]
+class Queue[A] protected (array: Array[AnyRef], start: Int, end: Int)
+  extends ArrayDeque[A](array, start, end)
+    with IndexedSeqOps[A, Queue, Queue[A]]
     with StrictOptimizedSeqOps[A, Queue, Queue[A]]
-    with Builder[A, Queue[A]]
     with Cloneable[Queue[A]]
-    with Serializable
-{
-  private[mutable] def this(fst: LinkedList[A], lst: LinkedList[A], lng: Int) = {
-    this()
-    first0 = fst
-    last0 = lst
-    len = lng
-  }
+    with Serializable {
+
+  def this(initialSize: Int = ArrayDeque.DefaultInitialSize) =
+    this(ArrayDeque.alloc(initialSize), start = 0, end = 0)
 
   override def iterableFactory: SeqFactory[Queue] = Queue
-  override protected[this] def fromSpecificIterable(coll: strawman.collection.Iterable[A]): Queue[A] = iterableFactory.from(coll)
-  override protected[this] def newSpecificBuilder(): Builder[A, Queue[A]] = iterableFactory.newBuilder()
 
-  /** Adds all elements to the queue.
+  /**
+    * Add elements to the end of this queue
     *
-    *  @param  elems       the elements to add.
+    * @param elem
+    * @return this
     */
-  def enqueue(elems: A*): Unit = this ++= elems.toStrawman
+  def enqueue(elem: A): this.type = this += elem
 
-  /** Returns the first element in the queue, and removes this element
-    *  from the queue.
+  /** Enqueue two or more elements at the end of the queue. The last element
+    *  of the sequence will be on end of the queue.
     *
-    *  @throws java.util.NoSuchElementException
-    *  @return the first element of the queue.
+    *  @param   elems      the element sequence.
+    *  @return this
     */
-  def dequeue(): A =
-    if (isEmpty)
-      throw new NoSuchElementException("queue empty")
-    else {
-      val res = first0.elem
-      first0 = first0.next
-      decrementLength()
-      res
-    }
+  def enqueue(elem1: A, elem2: A, elems: A*): this.type = enqueue(elem1).enqueue(elem2).enqueueAll(elems.toStrawman)
```

Comment:
This could force a resize to the correct target size before adding any elements if `elems.knownSize >= 0`